perception/radar_object_clustering/src/radar_object_clustering_node/radar_object_clustering_node.cpp

```diff
@@ -0,0 +1,220 @@
+
+#include "perception_utils/perception_utils.hpp"
+
+#include <tf2/utils.h>
+
+#include <cmath>
+#include <memory>
+#include <string>
+#include <vector>
+
+#ifdef ROS_DISTRO_GALACTIC
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#else
+#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
+#endif
+
+namespace
+{
+template <class T>
+bool update_param(
+  const std::vector<rclcpp::Parameter> & params, const std::string & name, T & value)
+{
+  const auto itr = std::find_if(
+    params.cbegin(), params.cend(),
+    [&name](const rclcpp::Parameter & p) { return p.get_name() == name; });
+
+  // Not found
+  if (itr == params.cend()) {
+    return false;
+  }
+
+  value = itr->template get_value<T>();
+  return true;
+}
+}  // namespace
+
+namespace radar_object_clustering
+{
+using autoware_auto_perception_msgs::msg::DetectedObject;
+using autoware_auto_perception_msgs::msg::DetectedObjects;
+
+RadarObjectClusteringNode::RadarObjectClusteringNode(const rclcpp::NodeOptions & node_options)
+: Node("radar_object_clustering", node_options)
+{
+  // Parameter Server
+  set_param_res_ = this->add_on_set_parameters_callback(
+    std::bind(&RadarObjectClusteringNode::onSetParam, this, std::placeholders::_1));
+
+  // Node Parameter
+  node_param_.angle_threshold = declare_parameter<double>("angle_threshold", 0.174);
+  node_param_.distance_threshold = declare_parameter<double>("distance_threshold", 4.0);
+  node_param_.velocity_threshold = declare_parameter<double>("velocity_threshold", 2.0);
+  node_param_.is_fixed_label = declare_parameter<bool>("is_fixed_label", false);
+  node_param_.fixed_label = declare_parameter<std::string>("fixed_label", "UNKNOWN");
+  node_param_.is_fixed_size = declare_parameter<bool>("is_fixed_size", false);
+  node_param_.size_x = declare_parameter<double>("size_x", 4.0);
+  node_param_.size_y = declare_parameter<double>("size_y", 1.5);
+  node_param_.size_z = declare_parameter<double>("size_z", 1.5);
+
+  // Subscriber
+  sub_objects_ = create_subscription<DetectedObjects>(
+    "~/input/objects", rclcpp::QoS{1},
+    std::bind(&RadarObjectClusteringNode::onObjects, this, std::placeholders::_1));
+
+  // Publisher
+  pub_objects_ = create_publisher<DetectedObjects>("~/output/objects", 1);
+}
+
+void RadarObjectClusteringNode::onObjects(const DetectedObjects::ConstSharedPtr objects_data_)
+{
+  DetectedObjects output_objects;
+  output_objects.header = objects_data_->header;
+  const auto & objects = objects_data_->objects;
+  std::vector<bool> used_flags(objects.size(), false);
+
+  for (size_t i = 0; i < objects.size(); i++) {
+    if (used_flags.at(i) == true) {
+      continue;
+    }
+
+    std::vector<DetectedObject> clustered_objects;
+    used_flags.at(i) = true;
+    clustered_objects.emplace_back(objects.at(i));
+
+    for (size_t j = i; j < objects.size(); ++j) {
+      if (used_flags.at(j) == false && isSameObject(objects.at(i), objects.at(j))) {
```

Comment:
This logic may affected by objects order.
How do you think about do matching between `objects.at(j)` and all `clustered_objects`?

Reply:
I added distance sort to prevent the result from changing depending on the order of objects in DetectedObjects at 50eb8b.
In addition, I added note that radar characteristics affect matching parameters to README at 7acc131.